internal/commands/build/build.go

```diff
@@ -0,0 +1,240 @@
+	out      string
+}
+
+func Cmd(dockerCli command.Cli) *cobra.Command {
+	var opts buildOptions
+	cmd := &cobra.Command{
+		Use:     "build [APPLICATION]",
+		Short:   "Build service images for the application",
+		Example: `$ docker app build myapp.dockerapp`,
+		Args:    cli.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			tag, err := runBuild(dockerCli, args[0], opts)
+			if err == nil {
+				fmt.Printf("Successfully build %s\n", tag.String())
+			}
+			return err
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVar(&opts.noCache, "no-cache", false, "Do not use cache when building the image")
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+	flags.StringVarP(&opts.out, "output", "o", "", "Dump generated bundle into a file")
+	flags.StringVarP(&opts.tag, "tag", "t", "", "Name and optionally a tag in the 'name:tag' format")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	err := checkMinimalEngineVersion(dockerCli)
+	if err != nil {
+		return nil, err
+	}
+
+	var ref reference.Named
+	ref, err = packager.GetNamedTagged(opt.tag)
+	if err != nil {
+		return nil, err
+	}
+
+	app, err := packager.Extract(application)
+	if err != nil {
+		return nil, err
+	}
+	defer app.Cleanup()
+
+	bundle, err := packager.MakeBundleFromApp(dockerCli, app, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts, err := parseCompose(app, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts["invocation-image"], err = createInvocationImageBuildOptions(dockerCli, app)
+	if err != nil {
+		return nil, err
+	}
+
+	debugBuildOpts(buildopts)
+
+	ctx, cancel := context.WithCancel(appcontext.Context())
+	defer cancel()
+	d, err := driver.GetDriver(ctx, "buildx_buildkit_default", nil, dockerCli.Client(), nil, "", nil)
+	if err != nil {
+		return nil, err
+	}
+	driverInfo := []build.DriverInfo{
+		{
+			Name:   "default",
+			Driver: d,
+		},
+	}
+
+	pw := progress.NewPrinter(ctx, os.Stderr, opt.progress)
+
+	// We rely on buildx "docker" builder integrated in docker engine, so don't nee a DockerAPI here
+	resp, err := build.Build(ctx, driverInfo, buildopts, nil, dockerCli.ConfigFile(), pw)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("Successfully built service images")
+	updateBundle(bundle, resp)
+
+	if ref == nil {
+		if ref, err = computeDigest(bundle); err != nil {
+			return nil, err
+		}
+	}
+
+	if err = persistBundle(opt, bundle, ref); err != nil {
+		return nil, err
+	}
+	return ref, nil
+}
+
+func checkMinimalEngineVersion(dockerCli command.Cli) error {
+	info, err := dockerCli.Client().Info(appcontext.Context())
+	if err != nil {
+		return err
+	}
+	majorVersion, err := strconv.Atoi(info.ServerVersion[:strings.IndexRune(info.ServerVersion, '.')])
```

Comment:
nit: `strings.SplitN(info.ServerVersion, ".", 2)[0]` should work too. `strings.IndexRune` returns `-1` if it can't find any dot...